src/ess/estia/calibration.py

```diff
@@ -0,0 +1,239 @@
+            * (rho * (Ipp - Ipa) - Iaa + Iap)
+        )
+    )
+    Ap = sc.sqrt(
+        (Ipp + Iaa - Ipa - Iap)
+        * (rho * (Ipp - Ipa) - Iaa + Iap)
+        / (
+            (rho * alp * Ipp + Iaa + rho * Ipa + alp * Iap)
+            * (alp * (Ipp - Iap) - Iaa + Ipa)
+        )
+    )
+    Rs = sc.sqrt(
+        (alp * (Ipp - Iap) - Iaa + Ipa)
+        * (rho * (Ipp - Ipa) - Iaa + Iap)
+        / ((rho * alp * Ipp + Iaa + rho * Ipa + alp * Iap) * (Ipp + Iaa - Ipa - Iap))
+    )
+
+    Pa = -rho * Pp
+    Aa = -alp * Ap
+
+    Rspp_minus_Rsaa = Rs * Rspp_plus_Rsaa
+    Rspp = (Rspp_plus_Rsaa + Rspp_minus_Rsaa) / 2
+    Rsaa = Rspp_plus_Rsaa - Rspp
+
+    return I0, Pp, Pa, Ap, Aa, Rspp, Rsaa
+
+
+def correction_matrix(Pp, Pa, Ap, Aa):
+    """
+    Defines the linear relationship between measured intensity
+    and reflectivity.
+    """
+    return [
+        [
+            (1 + Pp) * (1 + Ap) / 4,
+            (1 + Pp) * (1 - Ap) / 4,
+            (1 - Pp) * (1 + Ap) / 4,
+            (1 - Pp) * (1 - Ap) / 4,
+        ],
+        [
+            (1 + Pp) * (1 + Aa) / 4,
+            (1 + Pp) * (1 - Aa) / 4,
+            (1 - Pp) * (1 + Aa) / 4,
+            (1 - Pp) * (1 - Aa) / 4,
+        ],
+        [
+            (1 + Pa) * (1 + Ap) / 4,
+            (1 + Pa) * (1 - Ap) / 4,
+            (1 - Pa) * (1 + Ap) / 4,
+            (1 - Pa) * (1 - Ap) / 4,
+        ],
+        [
+            (1 + Pa) * (1 + Aa) / 4,
+            (1 + Pa) * (1 - Aa) / 4,
+            (1 - Pa) * (1 + Aa) / 4,
+            (1 - Pa) * (1 - Aa) / 4,
+        ],
+    ]
+
+
+def calibration_factors_from_reference_measurements(Io, Is):
+    """
+    Computes the polarization instrument parameters from
+    the calibration measurements on the non-magnetic reference
+    and the calibration measurements on the magnetic reference.
+    """
+    I0, Pp, Pa, Ap, Aa, _, _ = solve_for_calibration_parameters(Io, Is)
+    return I0, correction_matrix(Pp, Pa, Ap, Aa)
+
+
+def _linsolve(A, b):
+    x = np.linalg.solve(
+        np.stack([np.stack(row, -1) for row in A], -2),
+        np.stack(b, -1)[..., None],
+    )[..., 0]
+    return np.moveaxis(x, -1, 0)
+
+
+def linsolve(A, b):
+    x = _linsolve(
+        [[a.values for a in row] for row in A],
+        [bi.values for bi in b],
+    )
+    return [sc.array(dims=b[0].dims, values=xi) for xi in x]
+
+
+def compute_reflectivity_calibrate_on_q(
+    reference_supermirror,
+    reference_polarized_supermirror,
+    sample,
+    qbins,
+):
+    """Reduces reference and sample to Q before applying
+    the polarization correction and normalization."""
+    reference_supermirror = [reduce_to_q(i, qbins) for i in reference_supermirror]
+    reference_polarized_supermirror = [
+        reduce_to_q(i, qbins) for i in reference_polarized_supermirror
+    ]
+    I0, C = calibration_factors_from_reference_measurements(
+        reference_supermirror, reference_polarized_supermirror
+    )
```

Comment:
Here and in quite a few other functions: Please use (and enforce, if possible) function args to be keyword args. There are many simple mess-ups than can be avoided that way.